### ISSUE TITLE
Advanced generic call argument inference

### ISSUE THREAD

**Issue by ibraheemdev** [astral-sh/ty#2521]
We currently only account for the call expression type annotation when inferring the arguments of a generic call. This means we cannot currently solve calls where arguments are dependent on one another, e.g.,
```py
def lst[T](x: T) -> list[T]:
    return [x]

def f[T](x: T, y: list[T]) -> T:
    return x

def _(x: int, y: int | str):
    _: int | str = f(y, lst(x))
    f(y, lst(x)) # Argument to function `f` is incorrect: Expected `list[int | str]`, found `list[int]`
```

Pyright handles the above case fine, but mypy seems to give up. I suspect pyright is using a heuristic where it infers "simple" argument expressions first, i.e., arguments that cannot be influenced by type context. However, this means it is not able to solve more advanced cases, e.g.,
```py
def lst[T](x: T) -> list[T]:
    return [x]

def f[T](x: T, y: list[T], z: list[T]) -> T:
    return x

def _(x: int, y: int | str, z: int | str | None):
    _: int | str | None = f(y, lst(x), lst(z))
    f(y, lst(x), lst(z)) # Argument to function `f` is incorrect: Expected `list[int | str | None]`, found `list[int]`
```

Another problem is that we consider the call expression type annotation even if it is covariant position, which can lead to false positives, e.g.,
```py
from typing import Sequence

def lst[T](x: T) -> list[T]:
    return [x]

def f[T](x: T, y: list[T], z: list[T]) -> Sequence[T]:
    return [x]

def _(x: int, z: list[int]):
    _: Sequence[int] = f(x, lst(x), z)
    _: Sequence[int | str] = f(x, lst(x), z) # Argument to function `f` is incorrect: Expected `list[int | str]`, found `list[int]`
```

Another example with `TypedDict`:
```py
from typing import TypedDict

class A(TypedDict):
    a: int
    b: int

def lst[T](x: T) -> list[T]:
    return [x]

def f[T](x: T, y: list[T]) -> T:
    return x

def _(a: A):
    _: A = f(a, lst({ "a": 1, "b": 2 }))
    f(a, lst({ "a": 1, "b": 2 })) # Argument to function `f` is incorrect: Expected `list[B | dict[Unknown | str, Unknown | int]]`, found `list[dict[Unknown | str, Unknown | int]]`
```

We may be able to solve this problem completely by lazily inferring arguments as constraint sets, and unifying them after all arguments and type context has been inferred (cc @dcreager). We could also implement a simpler heuristic similar to pyright.

Note that some of these false positives are being hidden by our `Unknown` unioning, and might start popping up more frequently after https://github.com/astral-sh/ty/issues/1240 is resolved.

**Comment by ibraheemdev** [comment]
Another interesting example:
```py
class X:
    ...

def lst[T](x: T) -> list[T]:
    return [x]

def f[T](x: list[T | int], y: list[T | str]):
    ...

# Argument to function `f` is incorrect: Expected `list[X | int]`, found `list[X]`
# Argument to function `f` is incorrect: Expected `list[X | int]`, found `list[X]`
f(lst(X()), lst(X()))
```

This one should be relatively easy to fix.

**Comment by ibraheemdev** [comment]
After https://github.com/astral-sh/ruff/pull/23369 lands, we may be able to implement this without inferring constraint sets by instead propagating type context unconditionally. We may then infer types containing outer type variables (e.g., `list[T | int]`) for nested calls, and can unify by applying the outer specialization recursively to all nested inferred types. I believe this is equivalent to inferring constraint sets directly, but may be less invasive.

The `TypedDict` example may require a special dictionary literal types such that `DictLiteral[.., ..] | TD` simplifies to `TD` (but this is true of both approaches).

**Comment by ibraheemdev** [comment]
There's another interesting case now that we support lambda inference with type context:
```py
x = map(lambda x: reveal_type(x) + 2, [1, 2, 3]) # revealed: Unknown
reveal_type(x) # revealed: map[Unknown]
```

This one is a lot trickier to support, as we would need to support arbitrary operations on existential type variables if we were to infer the lambda expression without type context. On the other hand, uses within the lambda may enforce constraints on the type inferred for the list expression, so there is an unavoidable cycle here that is only solved by the constraint set approach (as opposed to inferring one with the type context of the other, which is not strictly correct).

**Comment by carljm** [comment]
Example from https://github.com/astral-sh/ty/issues/3575:

```py
from typing import reveal_type

class Item:
    pass

def sort_items(items: list[Item]) -> None:
    sorted(items, key=lambda item: reveal_type(item).nonexistent()) # reveals Unknown, no error about .nonexistent()
```

**Comment by ibraheemdev** [comment]
From https://github.com/astral-sh/ruff/pull/23186#issuecomment-4722551072, inferring constraint sets would also allow us to avoid literal promotion for arguments in nested generic calls, e.g.,
```py
def head[T](x: list[T]) -> T:
    return x[0]

reveal_type(head([1]))   # ideally: Literal[1]
```

Note that this would not be possible with the generic type context approach, as while we could also avoid promotion based on the presence of a type variable in `list[T]` which may resolve to a literal type, that would also mean avoiding promotion in cases where promotion is desirable, e.g.,
```py
def noop[T](x: list[T]) -> list[T]:
    return x

reveal_type(noop([1]))   # ideally: list[int]
```

**Comment by ibraheemdev** [comment]
Interestingly, pyright [seems to treat generic call inference as fixed point iteration](https://github.com/microsoft/pyright/blob/de5220a32dc46997d0d7940633cdcf963e443a77/packages/pyright-internal/src/analyzer/typeEvaluator.ts#L12234-L12275), using the lower bounds of type variables inferred from earlier arguments, in source-order, to specialize type context for later arguments, across multiple iterations. There is a maximum of three iterations, which means pyright fails to converge on pathological cases such as the following:

```py
from typing import Callable

def f[A, B, C](
    first: Callable[[C], C],
    second: Callable[[B], C],
    third: Callable[[A], B],
    source: list[A],
): ...

f(
    lambda c: reveal_type(c),      # revealed: Unknown
    lambda b: reveal_type(b) + 1,  # revealed: int
    lambda a: reveal_type(a) + 1,  # revealed: int
    reveal_type([1, 2, 3]),        # revealed: list[int]
)
```

Because pyright applies constraints in source-order, as opposed to collecting constraints from passes independently, it fails on examples like:
```py
def lst[T](x: T) -> list[T]:
    return [x]

def f[T](x: T, y: list[T], z: list[T]) -> T:
    return x

def _(x: int, y: int | str, z: int | str | None):
    f(y, lst(x), lst(z)) # pyright: Argument of type "list[int | str | None]" cannot be assigned to parameter "z" of type "list[T@f]" in function "f"
```

mypy implements using the heuristic I mentioned, [deferring complex expressions such as callables](https://github.com/python/mypy/blob/v2.0.0/mypy/checkexpr.py#L2237-L2296) to a second-pass, meaning it is able to solve the example above, but fails to solve other examples where expressions on the second-pass contribute further lower bound constraints:
```py
from typing import Callable

def f[T](producer: Callable[[], T], value: T) -> T:
    return value

result = f(lambda: "s", 1)  # mypy: error: Argument 1 to "choose" has incompatible type
```

Treating the lower bounds inferred from the first-pass as upper bound for the second pass is incorrect, though may work fine as a heuristic in practice, and is likely less expensive than pyright's approach.

**Comment by ibraheemdev** [comment]
https://github.com/astral-sh/ruff/pull/26141 (which is yet to land) should largely resolve this issue. I opened up https://github.com/astral-sh/ty/issues/3934 and https://github.com/astral-sh/ty/issues/3933 to track the remaining limitations separately, and https://github.com/astral-sh/ty/issues/2280 is already tracking the literal promotion improvement.
